src/kafunk/Compression.fs

```diff
@@ -67,81 +57,114 @@ module Snappy =
   
   open System
   open Snappy
-
-  module CompressedMessage = 
     
-    type BinaryZipper with
+  module internal Binary = 
+
+    let writeBlock (bytes: Binary.Segment) (buf : Binary.Segment) = 
+      Buffer.BlockCopy(bytes.Array, bytes.Offset, buf.Array, buf.Offset, bytes.Count)
+      Binary.shiftOffset bytes.Count buf
+
+    let readBlock (length: int) (buf : Binary.Segment) = 
+      let arr = ArraySegment<byte>(buf.Array, buf.Offset, length)
+      arr, Binary.shiftOffset length buf
+
+    let truncateIfSmaller actualLength maxLength (array: byte []) = 
```

Comment:
Yeah looks good. Also here:

```fsharp
let buf = Array.zeroCreate uncompressedLength
let actualLength = SnappyCodec.Uncompress(content.Array, content.Offset, content.Count, buf, 0)
Binary.truncateIfSmaller actualLength uncompressedLength buf
```

The call to `Binary.truncateIfSmaller` should operate on `ArraySegment<byte>` rather than array directly. This way, when it actually has to truncate, it doesn't call `Binary.toArray` which creates a new array, but simply adjusts the offsets on the ArraySegment.